controller/cmd/service-mirror/probe_worker.go

```diff
@@ -0,0 +1,140 @@
+package servicemirror
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	logging "github.com/sirupsen/logrus"
+)
+
+const httpGatewayTimeoutMillis = 50000
+
+// ProbeWorker is responsible for monitoring gateways using a probe specification
+type ProbeWorker struct {
+	*sync.RWMutex
+	probe          *GatewayProbeSpec
+	pairedServices map[string]bool
+	stopCh         chan struct{}
+	metrics        *probeMetrics
+	log            *logging.Entry
+}
+
+// NewProbeWorker creates a new probe worker associated with a particular gateway
+func NewProbeWorker(probe *GatewayProbeSpec, metrics *probeMetrics, probekey string) *ProbeWorker {
+	return &ProbeWorker{
+		RWMutex:        &sync.RWMutex{},
+		probe:          probe,
+		pairedServices: make(map[string]bool),
+		stopCh:         make(chan struct{}),
+		metrics:        metrics,
+		log: logging.WithFields(logging.Fields{
+			"probe-key": probekey,
+		}),
+	}
+}
+
+// NumPairedServices returns the number of paired services for this probe worker
+func (pw *ProbeWorker) NumPairedServices() int {
+	return len(pw.pairedServices)
+}
+
+// PairService increments the number of services that are routed by the gateway
+func (pw *ProbeWorker) PairService(serviceName, serviceNamespace string) {
+	svcKey := fmt.Sprintf("%s-%s", serviceNamespace, serviceName)
+	if _, ok := pw.pairedServices[svcKey]; !ok {
+		pw.pairedServices[svcKey] = true
+		pw.metrics.services.Inc()
+	}
+}
+
+// UnPairService decrements the number of services that are routed by the gateway
+func (pw *ProbeWorker) UnPairService(serviceName, serviceNamespace string) {
+	svcKey := fmt.Sprintf("%s-%s", serviceNamespace, serviceName)
+	if _, ok := pw.pairedServices[svcKey]; ok {
+		delete(pw.pairedServices, svcKey)
+		pw.metrics.services.Dec()
+	}
+}
+
+// UpdateProbeSpec is used to update the probe specification when something about the gateway changes
+func (pw *ProbeWorker) UpdateProbeSpec(spec *GatewayProbeSpec) {
+	pw.Lock()
+	pw.probe = spec
+	pw.Unlock()
+}
+
+// Stop this probe worker
+func (pw *ProbeWorker) Stop() {
+	pw.log.Debug("Stopping probe worker for")
+	close(pw.stopCh)
+}
+
+// Start this probe worker
+func (pw *ProbeWorker) Start() {
+	pw.log.Debug("Starting probe worker")
+	go pw.run()
+}
+
+func (pw *ProbeWorker) run() {
+	probeTickerPeriod := time.Duration(pw.probe.periodSeconds) * time.Second
+
+	// Introduce some randomness to avoid bursts of probes
+	time.Sleep(time.Duration(rand.Float64() * float64(probeTickerPeriod)))
```

Comment:
Using time.Sleep and Ticker together is a bit confusing.  Also, this can sleep up to `probeTickerPeriod` making the time between probes potentially as long as 2*`probeTickerPeriod`.  

Consider instead a ticker with some random jitter like https://github.com/projectcalico/felix/blob/master/jitter/jittered_ticker.go

Reply:
Makes sense, however I think this only reflects the start point. The time between probes will be equal to `probeTickerPeriod`. It just that there is delay of when the probing starts.

Reply:
But fine, I will just grab this jittered ticker. Looks like a nice utility anyway.